{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "import itertools\n",
    "import matplotlib_venn\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Load blast hits , hmmer hits and cluster stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(108, 11)\n",
      "(51, 9)\n"
     ]
    }
   ],
   "source": [
    "#Load blast best hits\n",
    "filt_blast_hits = pd.read_csv(\"1_out/filtered_blast_best_hits.csv\")\n",
    "print(filt_blast_hits.shape)\n",
    "#Load hmmer best hits\n",
    "filt_hmmer_hits = pd.read_csv(\"1_out/filtered_hmmer_best_hits.csv\")\n",
    "print(filt_hmmer_hits.shape)\n",
    "\n",
    "all_filt_hits = pd.concat([filt_blast_hits,filt_hmmer_hits],axis=0)\n",
    "#Load cluster stats\n",
    "clusters = pd.read_csv(\"1_out/cluster_stats.csv\")\n",
    "clusters[\"sample_origin\"] = clusters[\"sample_origin\"].apply(lambda x:x.replace(\",\",\";\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(159, 12)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>best_dmn_e_value</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>cluster</th>\n",
       "      <th>db</th>\n",
       "      <th>e_value</th>\n",
       "      <th>pct_id</th>\n",
       "      <th>q_cov</th>\n",
       "      <th>q_len</th>\n",
       "      <th>query_id</th>\n",
       "      <th>s_description</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>tool</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>488.0</td>\n",
       "      <td>1073</td>\n",
       "      <td>metahit_2014_cds</td>\n",
       "      <td>1.000000e-135</td>\n",
       "      <td>98.560</td>\n",
       "      <td>100.0</td>\n",
       "      <td>276.0</td>\n",
       "      <td>GB3LKKR01A150I</td>\n",
       "      <td>N074A_GL0017322  [gene]  locus=scaffold30187_1...</td>\n",
       "      <td>N074A_GL0017322</td>\n",
       "      <td>blastn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>479.0</td>\n",
       "      <td>1073</td>\n",
       "      <td>env_nt</td>\n",
       "      <td>1.000000e-131</td>\n",
       "      <td>98.180</td>\n",
       "      <td>99.0</td>\n",
       "      <td>276.0</td>\n",
       "      <td>GB3LKKR01A150I</td>\n",
       "      <td>gut metagenome genome assembly P2E0-k21-2014-0...</td>\n",
       "      <td>gi|936108378|emb|CEAB01076172.1|</td>\n",
       "      <td>blastn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>444.0</td>\n",
       "      <td>1073</td>\n",
       "      <td>metahit_cds</td>\n",
       "      <td>1.010000e-122</td>\n",
       "      <td>95.971</td>\n",
       "      <td>96.0</td>\n",
       "      <td>285.0</td>\n",
       "      <td>contig07331</td>\n",
       "      <td>GL0080651_MH0011_[Complete]_[mRNA]_locus=scaff...</td>\n",
       "      <td>GL0080651_MH0011_[Complete]_[mRNA]_locus=scaff...</td>\n",
       "      <td>blastn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>159.0</td>\n",
       "      <td>1073</td>\n",
       "      <td>metahit_pep</td>\n",
       "      <td>3.240000e-48</td>\n",
       "      <td>97.727</td>\n",
       "      <td>96.0</td>\n",
       "      <td>92.0</td>\n",
       "      <td>contig07331</td>\n",
       "      <td>GL0080651_MH0011_[Complete]_[mRNA]_locus=scaff...</td>\n",
       "      <td>GL0080651_MH0011_[Complete]_[mRNA]_locus=scaff...</td>\n",
       "      <td>blastp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>133.0</td>\n",
       "      <td>1073</td>\n",
       "      <td>metahit_2014_pep</td>\n",
       "      <td>8.000000e-38</td>\n",
       "      <td>100.000</td>\n",
       "      <td>84.0</td>\n",
       "      <td>82.0</td>\n",
       "      <td>GB3LKKR01CLC55</td>\n",
       "      <td>V1.CD14-0_GL0044222  [gene]  locus=scaffold182...</td>\n",
       "      <td>V1.CD14-0_GL0044222</td>\n",
       "      <td>blastp</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   best_dmn_e_value  bitscore cluster                db        e_value  \\\n",
       "0               NaN     488.0    1073  metahit_2014_cds  1.000000e-135   \n",
       "1               NaN     479.0    1073            env_nt  1.000000e-131   \n",
       "2               NaN     444.0    1073       metahit_cds  1.010000e-122   \n",
       "3               NaN     159.0    1073       metahit_pep   3.240000e-48   \n",
       "4               NaN     133.0    1073  metahit_2014_pep   8.000000e-38   \n",
       "\n",
       "    pct_id  q_cov  q_len        query_id  \\\n",
       "0   98.560  100.0  276.0  GB3LKKR01A150I   \n",
       "1   98.180   99.0  276.0  GB3LKKR01A150I   \n",
       "2   95.971   96.0  285.0     contig07331   \n",
       "3   97.727   96.0   92.0     contig07331   \n",
       "4  100.000   84.0   82.0  GB3LKKR01CLC55   \n",
       "\n",
       "                                       s_description  \\\n",
       "0  N074A_GL0017322  [gene]  locus=scaffold30187_1...   \n",
       "1  gut metagenome genome assembly P2E0-k21-2014-0...   \n",
       "2  GL0080651_MH0011_[Complete]_[mRNA]_locus=scaff...   \n",
       "3  GL0080651_MH0011_[Complete]_[mRNA]_locus=scaff...   \n",
       "4  V1.CD14-0_GL0044222  [gene]  locus=scaffold182...   \n",
       "\n",
       "                                          subject_id    tool  \n",
       "0                                    N074A_GL0017322  blastn  \n",
       "1                   gi|936108378|emb|CEAB01076172.1|  blastn  \n",
       "2  GL0080651_MH0011_[Complete]_[mRNA]_locus=scaff...  blastn  \n",
       "3  GL0080651_MH0011_[Complete]_[mRNA]_locus=scaff...  blastp  \n",
       "4                                V1.CD14-0_GL0044222  blastp  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(all_filt_hits.shape)\n",
    "all_filt_hits.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(32, 23)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cluster</th>\n",
       "      <th># nr seqs</th>\n",
       "      <th># raw seqs</th>\n",
       "      <th>Score</th>\n",
       "      <th>log RNAcode score</th>\n",
       "      <th>P</th>\n",
       "      <th>ORF length (aa)</th>\n",
       "      <th>log ORF length</th>\n",
       "      <th>complexity</th>\n",
       "      <th>min</th>\n",
       "      <th>...</th>\n",
       "      <th>min.1</th>\n",
       "      <th>avg.1</th>\n",
       "      <th>max.1</th>\n",
       "      <th># seqs with max length</th>\n",
       "      <th>composite score</th>\n",
       "      <th>DNA alignment</th>\n",
       "      <th>protein alignment</th>\n",
       "      <th>RNAcode figure</th>\n",
       "      <th>cluster_id</th>\n",
       "      <th>sample_origin</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1217</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>41.974</td>\n",
       "      <td>1.62</td>\n",
       "      <td>7.704e–10</td>\n",
       "      <td>91</td>\n",
       "      <td>1.96</td>\n",
       "      <td>hi</td>\n",
       "      <td>3.79</td>\n",
       "      <td>...</td>\n",
       "      <td>38</td>\n",
       "      <td>73.67</td>\n",
       "      <td>87</td>\n",
       "      <td>2</td>\n",
       "      <td>7.57</td>\n",
       "      <td>&lt;a href=\"all_nt_aln/cluster1217_DNA.aln.txt\"&gt;D...</td>\n",
       "      <td>&lt;a href=\"all_aa_aln/cluster1217_AA.aln.txt\"&gt;AA...</td>\n",
       "      <td>&lt;a href=\"rnacode_all/cluster1217_eps/hss-0.eps...</td>\n",
       "      <td>1217</td>\n",
       "      <td>feces</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>339b</td>\n",
       "      <td>7</td>\n",
       "      <td>11</td>\n",
       "      <td>33.153</td>\n",
       "      <td>1.52</td>\n",
       "      <td>1.069e–06</td>\n",
       "      <td>110</td>\n",
       "      <td>2.04</td>\n",
       "      <td>hi</td>\n",
       "      <td>3.89</td>\n",
       "      <td>...</td>\n",
       "      <td>63</td>\n",
       "      <td>79.14</td>\n",
       "      <td>106</td>\n",
       "      <td>2</td>\n",
       "      <td>7.52</td>\n",
       "      <td>&lt;a href=\"all_nt_aln/cluster339b_DNA.aln.txt\"&gt;D...</td>\n",
       "      <td>&lt;a href=\"all_aa_aln/cluster339b_AA.aln.txt\"&gt;AA...</td>\n",
       "      <td>&lt;a href=\"rnacode_all/cluster339b_eps/hss-0.eps...</td>\n",
       "      <td>339b</td>\n",
       "      <td>feces</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>211b</td>\n",
       "      <td>7</td>\n",
       "      <td>22</td>\n",
       "      <td>42.949</td>\n",
       "      <td>1.63</td>\n",
       "      <td>3.364e–14</td>\n",
       "      <td>80</td>\n",
       "      <td>1.90</td>\n",
       "      <td>hi</td>\n",
       "      <td>3.72</td>\n",
       "      <td>...</td>\n",
       "      <td>31</td>\n",
       "      <td>62.00</td>\n",
       "      <td>73</td>\n",
       "      <td>2</td>\n",
       "      <td>7.49</td>\n",
       "      <td>&lt;a href=\"all_nt_aln/cluster211b_DNA.aln.txt\"&gt;D...</td>\n",
       "      <td>&lt;a href=\"all_aa_aln/cluster211b_AA.aln.txt\"&gt;AA...</td>\n",
       "      <td>&lt;a href=\"rnacode_all/cluster211b_eps/hss-0.eps...</td>\n",
       "      <td>211b</td>\n",
       "      <td>feces</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>562</td>\n",
       "      <td>7</td>\n",
       "      <td>9</td>\n",
       "      <td>39.417</td>\n",
       "      <td>1.60</td>\n",
       "      <td>4.592e–07</td>\n",
       "      <td>61</td>\n",
       "      <td>1.79</td>\n",
       "      <td>hi</td>\n",
       "      <td>3.78</td>\n",
       "      <td>...</td>\n",
       "      <td>29</td>\n",
       "      <td>49.57</td>\n",
       "      <td>59</td>\n",
       "      <td>8</td>\n",
       "      <td>7.39</td>\n",
       "      <td>&lt;a href=\"all_nt_aln/cluster562_DNA.aln.txt\"&gt;DN...</td>\n",
       "      <td>&lt;a href=\"all_aa_aln/cluster562_AA.aln.txt\"&gt;AA&lt;/a&gt;</td>\n",
       "      <td>&lt;a href=\"rnacode_all/cluster562_eps/hss-0.eps\"...</td>\n",
       "      <td>562</td>\n",
       "      <td>CSF;serum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>297a</td>\n",
       "      <td>5</td>\n",
       "      <td>14</td>\n",
       "      <td>26.891</td>\n",
       "      <td>1.43</td>\n",
       "      <td>0.001</td>\n",
       "      <td>106</td>\n",
       "      <td>2.03</td>\n",
       "      <td>hi</td>\n",
       "      <td>3.90</td>\n",
       "      <td>...</td>\n",
       "      <td>59</td>\n",
       "      <td>96.60</td>\n",
       "      <td>106</td>\n",
       "      <td>8</td>\n",
       "      <td>7.38</td>\n",
       "      <td>&lt;a href=\"all_nt_aln/cluster297a_DNA.aln.txt\"&gt;D...</td>\n",
       "      <td>&lt;a href=\"all_aa_aln/cluster297a_AA.aln.txt\"&gt;AA...</td>\n",
       "      <td>&lt;a href=\"rnacode_all/cluster297a_eps/hss-0.eps...</td>\n",
       "      <td>297a</td>\n",
       "      <td>feces</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Cluster  # nr seqs  # raw seqs   Score  log RNAcode score          P  \\\n",
       "0    1217          6           6  41.974               1.62  7.704e–10   \n",
       "1    339b          7          11  33.153               1.52  1.069e–06   \n",
       "2    211b          7          22  42.949               1.63  3.364e–14   \n",
       "3     562          7           9  39.417               1.60  4.592e–07   \n",
       "4    297a          5          14  26.891               1.43      0.001   \n",
       "\n",
       "   ORF length (aa)  log ORF length complexity   min      ...       min.1  \\\n",
       "0               91            1.96         hi  3.79      ...          38   \n",
       "1              110            2.04         hi  3.89      ...          63   \n",
       "2               80            1.90         hi  3.72      ...          31   \n",
       "3               61            1.79         hi  3.78      ...          29   \n",
       "4              106            2.03         hi  3.90      ...          59   \n",
       "\n",
       "   avg.1  max.1  # seqs with max length  composite score  \\\n",
       "0  73.67     87                       2             7.57   \n",
       "1  79.14    106                       2             7.52   \n",
       "2  62.00     73                       2             7.49   \n",
       "3  49.57     59                       8             7.39   \n",
       "4  96.60    106                       8             7.38   \n",
       "\n",
       "                                       DNA alignment  \\\n",
       "0  <a href=\"all_nt_aln/cluster1217_DNA.aln.txt\">D...   \n",
       "1  <a href=\"all_nt_aln/cluster339b_DNA.aln.txt\">D...   \n",
       "2  <a href=\"all_nt_aln/cluster211b_DNA.aln.txt\">D...   \n",
       "3  <a href=\"all_nt_aln/cluster562_DNA.aln.txt\">DN...   \n",
       "4  <a href=\"all_nt_aln/cluster297a_DNA.aln.txt\">D...   \n",
       "\n",
       "                                   protein alignment  \\\n",
       "0  <a href=\"all_aa_aln/cluster1217_AA.aln.txt\">AA...   \n",
       "1  <a href=\"all_aa_aln/cluster339b_AA.aln.txt\">AA...   \n",
       "2  <a href=\"all_aa_aln/cluster211b_AA.aln.txt\">AA...   \n",
       "3  <a href=\"all_aa_aln/cluster562_AA.aln.txt\">AA</a>   \n",
       "4  <a href=\"all_aa_aln/cluster297a_AA.aln.txt\">AA...   \n",
       "\n",
       "                                      RNAcode figure cluster_id sample_origin  \n",
       "0  <a href=\"rnacode_all/cluster1217_eps/hss-0.eps...       1217         feces  \n",
       "1  <a href=\"rnacode_all/cluster339b_eps/hss-0.eps...       339b         feces  \n",
       "2  <a href=\"rnacode_all/cluster211b_eps/hss-0.eps...       211b         feces  \n",
       "3  <a href=\"rnacode_all/cluster562_eps/hss-0.eps\"...        562     CSF;serum  \n",
       "4  <a href=\"rnacode_all/cluster297a_eps/hss-0.eps...       297a         feces  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(clusters.shape)\n",
    "clusters.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Summary of db hits per cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cluster families with no annotation: 3\n",
      "(32, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cluster</th>\n",
       "      <th>db_hit_count</th>\n",
       "      <th>dbs</th>\n",
       "      <th>sample_origin</th>\n",
       "      <th>composite score</th>\n",
       "      <th>P</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1217</td>\n",
       "      <td>7.0</td>\n",
       "      <td>metahit_cds;env_nt;metahit_2014_cds;metahit_20...</td>\n",
       "      <td>feces</td>\n",
       "      <td>7.57</td>\n",
       "      <td>7.704e–10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>339b</td>\n",
       "      <td>5.0</td>\n",
       "      <td>metahit_2014_cds;metahit_cds;env_nt;metahit_20...</td>\n",
       "      <td>feces</td>\n",
       "      <td>7.52</td>\n",
       "      <td>1.069e–06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>211b</td>\n",
       "      <td>5.0</td>\n",
       "      <td>env_nt;metahit_2014_cds;metahit_cds;metahit_pe...</td>\n",
       "      <td>feces</td>\n",
       "      <td>7.49</td>\n",
       "      <td>3.364e–14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>562</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CSF;serum</td>\n",
       "      <td>7.39</td>\n",
       "      <td>4.592e–07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>297a</td>\n",
       "      <td>5.0</td>\n",
       "      <td>metahit_2014_cds;metahit_cds;env_nt;metahit_20...</td>\n",
       "      <td>feces</td>\n",
       "      <td>7.38</td>\n",
       "      <td>0.001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Cluster  db_hit_count                                                dbs  \\\n",
       "3     1217           7.0  metahit_cds;env_nt;metahit_2014_cds;metahit_20...   \n",
       "16    339b           5.0  metahit_2014_cds;metahit_cds;env_nt;metahit_20...   \n",
       "9     211b           5.0  env_nt;metahit_2014_cds;metahit_cds;metahit_pe...   \n",
       "29     562           NaN                                                NaN   \n",
       "13    297a           5.0  metahit_2014_cds;metahit_cds;env_nt;metahit_20...   \n",
       "\n",
       "   sample_origin  composite score          P  \n",
       "3          feces             7.57  7.704e–10  \n",
       "16         feces             7.52  1.069e–06  \n",
       "9          feces             7.49  3.364e–14  \n",
       "29     CSF;serum             7.39  4.592e–07  \n",
       "13         feces             7.38      0.001  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmp = [(cluster,hits.shape[0],\";\".join(hits.db)) for cluster,hits in ( all_filt_hits[[\"cluster\",\"db\"]].drop_duplicates().groupby(\"cluster\") )]\n",
    "hits_x_cluster = pd.DataFrame.from_records(tmp,columns=[\"cluster\",\"db_hit_count\",\"dbs\"])\n",
    "print(\"Cluster families with no annotation: {}\".format(clusters.shape[0]-hits_x_cluster.shape[0]))\n",
    "hits_x_cluster = hits_x_cluster.merge(clusters,left_on=\"cluster\",right_on=\"Cluster\",how=\"right\")[[\"Cluster\",\"db_hit_count\",\"dbs\",\"sample_origin\",\"composite score\",\"P\"]]\n",
    "print(hits_x_cluster.shape)\n",
    "hits_x_cluster = hits_x_cluster.sort_values(by=\"composite score\",ascending=False)\n",
    "hits_x_cluster.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "hits_x_cluster.to_csv(\"2_out/dbhits_x_cluster.csv\",index=False,quote=\"\\\"\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. Summary of db hits grouped by sample origin"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Hits per db per sample_origin"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_nr</th>\n",
       "      <th>env_nt</th>\n",
       "      <th>metahit_2014_cds</th>\n",
       "      <th>metahit_2014_pep</th>\n",
       "      <th>metahit_cds</th>\n",
       "      <th>metahit_pep</th>\n",
       "      <th>nr</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_origin</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CSF;feces</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CSF;mucus;serum</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CSF;serum</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>feces</th>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>17</td>\n",
       "      <td>22</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>serum</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 env_nr  env_nt  metahit_2014_cds  metahit_2014_pep  \\\n",
       "sample_origin                                                         \n",
       "CSF;feces             0       1                 1                 1   \n",
       "CSF;mucus;serum       0       1                 0                 1   \n",
       "CSF;serum             0       1                 0                 0   \n",
       "feces                 1      23                22                22   \n",
       "serum                 0       0                 0                 0   \n",
       "\n",
       "                 metahit_cds  metahit_pep  nr  \n",
       "sample_origin                                  \n",
       "CSF;feces                  1            0   0  \n",
       "CSF;mucus;serum            0            0   0  \n",
       "CSF;serum                  0            0   1  \n",
       "feces                     17           22   2  \n",
       "serum                      0            0   1  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hits_x_sample_origin = {}\n",
    "for db in all_filt_hits.db.unique():\n",
    "    hits_x_sample_origin[db] = hits_x_cluster.groupby(\"sample_origin\").apply(lambda df: sum([(db in x.split(\";\")) if type(x) is str else False for x in df[\"dbs\"]]))\n",
    "hits_x_sample_origin = pd.concat(hits_x_sample_origin,axis=1)\n",
    "hits_x_sample_origin"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Calculate hit count to metahit = (metahit_pep or metahit_cds ), and env = ( env nr or env nt ) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_</th>\n",
       "      <th>metahit_</th>\n",
       "      <th>metahit_2014</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_origin</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CSF;feces</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CSF;mucus;serum</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CSF;serum</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>feces</th>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>serum</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 env_  metahit_  metahit_2014\n",
       "sample_origin                                \n",
       "CSF;feces           1         1             1\n",
       "CSF;mucus;serum     1         1             1\n",
       "CSF;serum           1         0             0\n",
       "feces              23        23            23\n",
       "serum               0         0             0"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hits_x_sample_origin_sum = {}\n",
    "for db in [\"env_\",\"metahit_\",\"metahit_2014\"]:\n",
    "    hits_x_sample_origin_sum[db] = hits_x_cluster.groupby(\"sample_origin\").apply(lambda df: sum([(db in x) if type(x) is str else False for x in df[\"dbs\"]]))\n",
    "hits_x_sample_origin_sum = pd.concat(hits_x_sample_origin_sum,axis=1)\n",
    "hits_x_sample_origin_sum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cluster</th>\n",
       "      <th>env_nr</th>\n",
       "      <th>env_nt</th>\n",
       "      <th>metahit_2014_cds</th>\n",
       "      <th>metahit_2014_pep</th>\n",
       "      <th>metahit_cds</th>\n",
       "      <th>metahit_pep</th>\n",
       "      <th>nr</th>\n",
       "      <th>env_</th>\n",
       "      <th>metahit_</th>\n",
       "      <th>metahit_2014</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_origin</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CSF;feces</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CSF;mucus;serum</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CSF;serum</th>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>feces</th>\n",
       "      <td>23</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>17</td>\n",
       "      <td>22</td>\n",
       "      <td>2</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>serum</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Cluster  env_nr  env_nt  metahit_2014_cds  metahit_2014_pep  \\\n",
       "sample_origin                                                                  \n",
       "CSF;feces              1       0       1                 1                 1   \n",
       "CSF;mucus;serum        2       0       1                 0                 1   \n",
       "CSF;serum              5       0       1                 0                 0   \n",
       "feces                 23       1      23                22                22   \n",
       "serum                  1       0       0                 0                 0   \n",
       "\n",
       "                 metahit_cds  metahit_pep  nr  env_  metahit_  metahit_2014  \n",
       "sample_origin                                                                \n",
       "CSF;feces                  1            0   0     1         1             1  \n",
       "CSF;mucus;serum            0            0   0     1         1             1  \n",
       "CSF;serum                  0            0   1     1         0             0  \n",
       "feces                     17           22   2    23        23            23  \n",
       "serum                      0            0   1     0         0             0  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table_2 = pd.concat([hits_x_cluster.groupby(\"sample_origin\").count()[\"Cluster\"],\n",
    "            hits_x_sample_origin,\n",
    "            hits_x_sample_origin_sum],axis=1)\n",
    "table_2.to_csv(\"2_out/dbhits_x_sample_origin.csv\")\n",
    "table_2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4. Analyze NCBI Environmental hits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(27, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>db</th>\n",
       "      <th>env_source</th>\n",
       "      <th>sample_origin</th>\n",
       "      <th>composite score</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cluster</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1073</th>\n",
       "      <td>1073</td>\n",
       "      <td>env_nt</td>\n",
       "      <td>gut</td>\n",
       "      <td>feces</td>\n",
       "      <td>7.19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113b</th>\n",
       "      <td>113b</td>\n",
       "      <td>env_nt</td>\n",
       "      <td>Gut</td>\n",
       "      <td>feces</td>\n",
       "      <td>6.32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1217</th>\n",
       "      <td>1217</td>\n",
       "      <td>env_nr</td>\n",
       "      <td>marine sediment</td>\n",
       "      <td>feces</td>\n",
       "      <td>7.57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1217</th>\n",
       "      <td>1217</td>\n",
       "      <td>env_nt</td>\n",
       "      <td>gut</td>\n",
       "      <td>feces</td>\n",
       "      <td>7.57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179a</th>\n",
       "      <td>179a</td>\n",
       "      <td>env_nt</td>\n",
       "      <td>gut</td>\n",
       "      <td>feces</td>\n",
       "      <td>7.23</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        cluster      db        env_source sample_origin  composite score\n",
       "cluster                                                                 \n",
       "1073       1073  env_nt              gut          feces             7.19\n",
       "113b       113b  env_nt              Gut          feces             6.32\n",
       "1217       1217  env_nr  marine sediment          feces             7.57\n",
       "1217       1217  env_nt              gut          feces             7.57\n",
       "179a       179a  env_nt              gut          feces             7.23"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Extract all env hits\n",
    "env_hits = all_filt_hits[[\"cluster\",\"db\",\"s_description\"]][all_filt_hits.db.apply(lambda x: \"env_\" in x)].sort_values(by=\"cluster\")\n",
    "#Parse description to extract the source metagenome\n",
    "env_hits[\"env_source\"] = env_hits.s_description.apply(lambda x: x.split(\"metagenome\")[0])\n",
    "env_hits[\"env_source\"] = env_hits[\"env_source\"].apply(lambda x: x.split(\"[\")[1] if \"[\" in x else x   )\n",
    "#Annotate clusters with sample origin\n",
    "env_hits = env_hits.merge(clusters[[\"Cluster\",\"sample_origin\",\"composite score\"]],left_on=\"cluster\",right_on=\"Cluster\")\n",
    "env_hits = env_hits.drop([\"s_description\",\"Cluster\"],axis=1)\n",
    "env_hits.index = env_hits[\"cluster\"]\n",
    "#Remove duplicates when blastp and hmmsearch hits are the same\n",
    "env_hits = env_hits.drop_duplicates()\n",
    "print(env_hits.shape)\n",
    "env_hits.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4.1 Environmental hits per cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cluster</th>\n",
       "      <th>composite score</th>\n",
       "      <th>sample_origin</th>\n",
       "      <th>env_hits</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1217</td>\n",
       "      <td>7.57</td>\n",
       "      <td>feces</td>\n",
       "      <td>marine sediment (p);gut (n)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>339b</td>\n",
       "      <td>7.52</td>\n",
       "      <td>feces</td>\n",
       "      <td>gut (n)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>211b</td>\n",
       "      <td>7.49</td>\n",
       "      <td>feces</td>\n",
       "      <td>gut (n)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>562</td>\n",
       "      <td>7.39</td>\n",
       "      <td>CSF;serum</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>297a</td>\n",
       "      <td>7.38</td>\n",
       "      <td>feces</td>\n",
       "      <td>gut (n)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Cluster  composite score sample_origin                     env_hits\n",
       "0    1217             7.57         feces  marine sediment (p);gut (n)\n",
       "1    339b             7.52         feces                      gut (n)\n",
       "2    211b             7.49         feces                      gut (n)\n",
       "3     562             7.39     CSF;serum                          NaN\n",
       "4    297a             7.38         feces                      gut (n)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmp = env_hits.groupby(\"cluster\").apply( lambda df: \";\".join( \"{}({})\".format(env_src,\"n\" if db == \"env_nt\" else \"p\") for _,db,env_src in df[[\"db\",\"env_source\"]].drop_duplicates().to_records()) )\n",
    "env_hits_x_clust = pd.merge(clusters[[\"Cluster\",\"composite score\",\"sample_origin\"]], \n",
    "                             pd.DataFrame(tmp,columns=[\"env_hits\"]),\n",
    "                             left_on=\"Cluster\",right_index=True,how=\"left\").sort_values(by=\"composite score\",ascending=False)\n",
    "env_hits_x_clust.to_csv(\"2_out/env_hits_x_cluster.csv\",index=False)\n",
    "env_hits_x_clust.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4.2 Environmental hits per sample origin"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_hits</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_origin</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CSF;feces</th>\n",
       "      <td>gut (1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CSF;mucus;serum</th>\n",
       "      <td>wastewater (1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CSF;serum</th>\n",
       "      <td>wastewater (1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>feces</th>\n",
       "      <td>gut (17); Human gut (5); marine sediment (1); ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                          env_hits\n",
       "sample_origin                                                     \n",
       "CSF;feces                                                  gut (1)\n",
       "CSF;mucus;serum                                     wastewater (1)\n",
       "CSF;serum                                           wastewater (1)\n",
       "feces            gut (17); Human gut (5); marine sediment (1); ..."
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "env_hits_x_sample_type = pd.DataFrame(env_hits.groupby(\"sample_origin\").apply(lambda df: \"; \".join(\"{}({})\".format(k,v) for k,v in df.groupby(\"env_source\").count()[\"cluster\"].to_dict().items())))\n",
    "env_hits_x_sample_type.columns = [\"env_hits\"]\n",
    "env_hits_x_sample_type.to_csv(\"2_out/env_hits_per_sample_origin.csv\")\n",
    "env_hits_x_sample_type"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 5. Summarize MetaHIT hits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cluster</th>\n",
       "      <th>composite score</th>\n",
       "      <th>sample_origin</th>\n",
       "      <th>MetaHIT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1217</td>\n",
       "      <td>7.57</td>\n",
       "      <td>feces</td>\n",
       "      <td>2014_n;2014_p;n;p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>339b</td>\n",
       "      <td>7.52</td>\n",
       "      <td>feces</td>\n",
       "      <td>2014_n;2014_p;n;p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>211b</td>\n",
       "      <td>7.49</td>\n",
       "      <td>feces</td>\n",
       "      <td>2014_n;2014_p;n;p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>562</td>\n",
       "      <td>7.39</td>\n",
       "      <td>CSF;serum</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>297a</td>\n",
       "      <td>7.38</td>\n",
       "      <td>feces</td>\n",
       "      <td>2014_n;2014_p;n;p</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Cluster  composite score sample_origin            MetaHIT\n",
       "0    1217             7.57         feces  2014_n;2014_p;n;p\n",
       "1    339b             7.52         feces  2014_n;2014_p;n;p\n",
       "2    211b             7.49         feces  2014_n;2014_p;n;p\n",
       "3     562             7.39     CSF;serum                NaN\n",
       "4    297a             7.38         feces  2014_n;2014_p;n;p"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metahit_hits = all_filt_hits[[\"cluster\",\"db\"]][all_filt_hits.db.apply(lambda x: \"metahit_\" in x)].sort_values(by=\"cluster\")\n",
    "metahit_x_cluster = metahit_hits.drop_duplicates().groupby(\"cluster\").apply(lambda df: \";\".join(sorted(df.db)))\n",
    "metahit_x_cluster= pd.DataFrame(metahit_x_cluster.apply(lambda x: x.replace(\"metahit_\",\"\").replace(\"cds\",\"n\").replace(\"pep\",\"p\")),columns=[\"MetaHIT\"])\n",
    "metahit_x_cluster = pd.merge(clusters[[\"Cluster\",\"composite score\",\"sample_origin\"]], \n",
    "                             metahit_x_cluster,\n",
    "                             left_on=\"Cluster\",right_index=True,how=\"left\")\n",
    "metahit_x_cluster.sort_values(by=\"composite score\",ascending=False).to_csv(\"2_out/metahit_hits_x_cluster.csv\",index=False)\n",
    "metahit_x_cluster.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 6. Hits per database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>db</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>env_nr</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>env_nt</th>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>metahit_2014_cds</th>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>metahit_2014_pep</th>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>metahit_cds</th>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>metahit_pep</th>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nr</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  cluster\n",
       "db                       \n",
       "env_nr                  1\n",
       "env_nt                 26\n",
       "metahit_2014_cds       23\n",
       "metahit_2014_pep       24\n",
       "metahit_cds            18\n",
       "metahit_pep            22\n",
       "nr                      4"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_filt_hits[[\"cluster\",\"db\"]].drop_duplicates().groupby(\"db\").count()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 7. NR/HMP hits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>db</th>\n",
       "      <th>tool</th>\n",
       "      <th>s_description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1217</td>\n",
       "      <td>nr</td>\n",
       "      <td>blastp</td>\n",
       "      <td>unknown [Veillonella sp. CAG:933]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1217</td>\n",
       "      <td>nr</td>\n",
       "      <td>hmmsearch</td>\n",
       "      <td>unknown [Veillonella sp. CAG:933]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>532</td>\n",
       "      <td>nr</td>\n",
       "      <td>hmmsearch</td>\n",
       "      <td>hypothetical protein [Megamonas rupellensis]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>565b</td>\n",
       "      <td>nr</td>\n",
       "      <td>blastp</td>\n",
       "      <td>hypothetical protein H257_12751 [Aphanomyces a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>565b</td>\n",
       "      <td>nr</td>\n",
       "      <td>hmmsearch</td>\n",
       "      <td>putative replication protein [Phytophthora par...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>956b</td>\n",
       "      <td>nr</td>\n",
       "      <td>blastp</td>\n",
       "      <td>hypothetical protein [Chlamydia trachomatis]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>956b</td>\n",
       "      <td>nr</td>\n",
       "      <td>hmmsearch</td>\n",
       "      <td>hypothetical protein [Chlamydia trachomatis]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    cluster  db       tool                                      s_description\n",
       "15     1217  nr     blastp                  unknown [Veillonella sp. CAG:933]\n",
       "7      1217  nr  hmmsearch                  unknown [Veillonella sp. CAG:933]\n",
       "42      532  nr  hmmsearch       hypothetical protein [Megamonas rupellensis]\n",
       "95     565b  nr     blastp  hypothetical protein H257_12751 [Aphanomyces a...\n",
       "43     565b  nr  hmmsearch  putative replication protein [Phytophthora par...\n",
       "107    956b  nr     blastp       hypothetical protein [Chlamydia trachomatis]\n",
       "50     956b  nr  hmmsearch       hypothetical protein [Chlamydia trachomatis]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nrhmp_hits = all_filt_hits[all_filt_hits.db.apply(lambda x: \"env\" not in x and \"metahit\" not in x )].sort_values(by=\"cluster\")\n",
    "nrhmp_hits = nrhmp_hits[[\"cluster\",\"db\",\"tool\",\"s_description\"]]\n",
    "nrhmp_hits.to_csv(\"2_out/nr_hmp_hits.csv\",index=False)\n",
    "nrhmp_hits"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Inspect individual hits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tool</th>\n",
       "      <th>db</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>pct_id</th>\n",
       "      <th>q_cov</th>\n",
       "      <th>q_len</th>\n",
       "      <th>s_description</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>best_dmn_e_value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hmmsearch</td>\n",
       "      <td>metahit_2014_pep</td>\n",
       "      <td>31.7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>strand:+  start:8853 stop:9821 length:969 star...</td>\n",
       "      <td>160218816-stool1_revised_C882654_1_gene76494</td>\n",
       "      <td>0.00015</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        tool                db  bitscore  pct_id  q_cov  q_len  \\\n",
       "2  hmmsearch  metahit_2014_pep      31.7     NaN    NaN    NaN   \n",
       "\n",
       "                                       s_description  \\\n",
       "2  strand:+  start:8853 stop:9821 length:969 star...   \n",
       "\n",
       "                                     subject_id  best_dmn_e_value  \n",
       "2  160218816-stool1_revised_C882654_1_gene76494           0.00015  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_filt_hits[all_filt_hits.cluster == \"1085\"][[\"tool\",\"db\",\"bitscore\",\"pct_id\",\"q_cov\",\"q_len\",\"s_description\",\"subject_id\",\"best_dmn_e_value\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "metahit_cds_list = set(all_filt_hits[all_filt_hits.db == \"metahit_cds\"][\"cluster\"].drop_duplicates().tolist())\n",
    "metahit_pep_list = set(all_filt_hits[all_filt_hits.db == \"metahit_pep\"][\"cluster\"].drop_duplicates().tolist())\n",
    "env_nt_list = set(all_filt_hits[all_filt_hits.db == \"env_nt\"][\"cluster\"].drop_duplicates().tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
